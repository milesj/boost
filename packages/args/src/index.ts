/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import format from './format';
import parse from './parse';
import parseInContext from './parseInContext';
import ArgsError from './ArgsError';
import type { ArgsErrorCode } from './ArgsError';
import ParseError from './ParseError';
import ValidationError from './ValidationError';

export * from './constants';
export * from './types';

export { format, parse, parseInContext, ArgsError, ParseError, ValidationError };
export type { ArgsErrorCode };
