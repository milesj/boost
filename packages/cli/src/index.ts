/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
