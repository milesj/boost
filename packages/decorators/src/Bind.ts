import isMethod from './helpers/isMethod';

export default function Bind(): MethodDecorator {
  return (target, property, descriptor) => {
    if (
      !isMethod(target, property, descriptor) ||
      !('value' in descriptor && typeof descriptor.value === 'function')
    ) {
      throw new TypeError(`\`@Bind\` may only be applied to class methods.`);
    }

    const func = descriptor.value;

    return {
      configurable: true,
      get(this: Function) {
        const bound = func.bind(this);

        // Only cache the bound function when in the deepest sub-class,
        // otherwise any `super` calls will overwrite each other.
        if (target.constructor.name === this.constructor.name) {
          Object.defineProperty(this, property, {
            configurable: true,
            value: bound,
            writable: true,
          });
        }

        return bound;
      },
    };
  };
}
