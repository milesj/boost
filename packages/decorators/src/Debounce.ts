import isMethod from './helpers/isMethod';

export default function Debounce(delay: number): MethodDecorator {
	return (target, property, descriptor) => {
		if (__DEV__ && (
				!isMethod(target, property, descriptor) ||
				!('value' in descriptor && typeof descriptor.value === 'function')
			)) {
				throw new TypeError(`\`@Debounce\` may only be applied to class methods.`);
			}

		// We must use a map as all class instances would share the
		// same timer value otherwise.
		const timers = new WeakMap<Function, NodeJS.Timeout>();

		// Overwrite the value function with a new debounced function
		const func = descriptor.value;

		// @ts-expect-error
		descriptor.value = function debounce(this: Function, ...args: unknown[]) {
			const timer = timers.get(this);

			if (timer) {
				clearTimeout(timer);
				timers.delete(this);
			}

			timers.set(
				this,
				setTimeout(() => {
					(func as unknown as Function).apply(this, args);
				}, delay),
			);
		};
	};
}
