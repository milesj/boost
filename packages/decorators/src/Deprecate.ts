/* eslint-disable no-console */

import isClass from './helpers/isClass';
import isMethod from './helpers/isMethod';
import isProperty from './helpers/isProperty';
import isParam from './helpers/isParam';

export default function Deprecate(message?: string) {
  return (target: Function | Object, property?: string | symbol, descriptor?: unknown): void => {
    const isProtoOrStatic = typeof target === 'function';
    const className = isProtoOrStatic ? (target as Function).name : target.constructor.name;
    const accessSymbol = isProtoOrStatic ? `.${String(property)}` : `#${String(property)}`;

    // Class
    if (isClass(target, property, descriptor)) {
      console.debug(message || `Class \`${className}\` has been deprecated.`);

      // Method
    } else if (isMethod(target, property, descriptor)) {
      console.debug(message || `Method \`${className + accessSymbol}()\` has been deprecated.`);

      // Property
    } else if (isProperty(target, property, descriptor)) {
      console.debug(message || `Property \`${className + accessSymbol}\` has been deprecated.`);

      // Param (Babel/Jest doesnt support them)
    } /* istanbul ignore next */ else if (isParam(target, property, descriptor)) {
      console.debug(
        message ||
          `Parameter ${descriptor} for \`${className + accessSymbol}()\` has been deprecated.`,
      );
    }
  };
}
