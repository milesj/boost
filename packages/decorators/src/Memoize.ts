/* eslint-disable babel/no-invalid-this */

import isMethod from './helpers/isMethod';

export type MemoizedFunction<T> = (...args: unknown[]) => T;

export type MemoizeHasher = (...args: unknown[]) => string;

export type MemoizeCache<T> = Map<
  string,
  {
    time?: number | null;
    value: T;
  }
>;

export interface MemoizeOptions<T> {
  cache?: MemoizeCache<T> | null;
  expires?: number;
  hasher?: MemoizeHasher;
}

function hasher(...args: unknown[]): string {
  return JSON.stringify(args);
}

function createMemoizer<T>(
  method: MemoizedFunction<T>,
  rootCache: WeakMap<Function, MemoizeCache<T>>,
  options: Required<MemoizeOptions<T>>,
): MemoizedFunction<T> {
  // Must be a normal function as we require `this`
  return function memoizer(this: Function, ...args: unknown[]) {
    // Extract the cache for this specific instance
    let cache = rootCache.get(this);

    if (!cache) {
      cache = options.cache ? new Map(options.cache) : new Map();
      rootCache.set(this, cache);
    }

    // Hash the key and check the cache
    const key = options.hasher(...args);
    const item = cache.get(key);

    if (item && (!item.time || (typeof item.time === 'number' && item.time > Date.now()))) {
      return item.value;
    }

    // No cache so execute and cache the result
    const value = method.apply(this, args);
    const time = options.expires > 0 ? Date.now() + options.expires : null;

    cache.set(key, {
      time,
      value,
    });

    // Only cache if successful
    if (value instanceof Promise) {
      value.catch(() => cache?.delete(key));
    }

    return value;
  };
}

export default function Memoize<T>(
  options: MemoizeHasher | MemoizeOptions<T> = {},
): MethodDecorator {
  // eslint-disable-next-line complexity
  return (target, property, descriptor) => {
    if (__DEV__) {
      if (
        !isMethod(target, property, descriptor) ||
        (!('value' in descriptor && typeof descriptor.value === 'function') &&
          !('get' in descriptor && typeof descriptor.get === 'function'))
      ) {
        throw new TypeError(`\`@Memoize\` may only be applied to class methods or getters.`);
      }
    }

    const config = {
      cache: null,
      expires: 0,
      hasher,
      ...(typeof options === 'function' ? { hasher: options } : options),
    };

    if (__DEV__) {
      if (config.cache && !(config.cache instanceof Map)) {
        throw new Error('`cache` must be an instance of `Map`.');
      }

      if (typeof config.expires !== 'number' || config.expires < 0) {
        throw new Error('`expires` must be a number greater than or equal to 0.');
      }

      if (typeof config.hasher !== 'function') {
        throw new TypeError('`hasher` must be a function.');
      }
    }

    // We must use a map as all class instances would share the
    // same cache otherwise. Probability of collision is high.
    const rootCache = new WeakMap<Function, MemoizeCache<T>>();

    if (descriptor.get) {
      // @ts-expect-error
      descriptor.get = createMemoizer<T>(descriptor.get, rootCache, config);
    } else if (descriptor.value) {
      // @ts-expect-error
      descriptor.value = createMemoizer<T>(descriptor.value, rootCache, config);
    }
  };
}
