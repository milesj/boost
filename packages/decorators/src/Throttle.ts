import { isMethod } from './helpers/isMethod';

export function Throttle(delay: number): MethodDecorator {
	return (target, property, descriptor) => {
		if (
			__DEV__ &&
			(!isMethod(target, property, descriptor) ||
				!('value' in descriptor && typeof descriptor.value === 'function'))
		) {
			throw new TypeError(`\`@Throttle\` may only be applied to class methods.`);
		}

		// We must use a map as all class instances would share the
		// same boolean value otherwise.
		const throttling = new WeakMap<Function, boolean>();

		// Overwrite the value function with a new throttled function
		const func = descriptor.value;

		// @ts-expect-error Override generic
		descriptor.value = function throttle(this: Function, ...args: unknown[]) {
			if (throttling.get(this)) {
				return;
			}

			(func as unknown as Function).apply(this, args);
			throttling.set(this, true);

			setTimeout(() => {
				throttling.delete(this);
			}, delay);
		};
	};
}
