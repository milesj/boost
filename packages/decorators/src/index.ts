/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
export { default as Memoize } from './Memoize';
export { default as Throttle } from './Throttle';
