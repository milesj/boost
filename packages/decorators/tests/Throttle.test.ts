import { afterEach, beforeEach, describe, expect, it, MockInstance, vi } from 'vitest';
import { Throttle } from '../src';

describe('@Throttle()', () => {
	class Test {
		@Throttle(100)
		log(...args: unknown[]) {
			console.log('Logged!', ...args);
		}
	}

	let spy: MockInstance;

	beforeEach(() => {
		vi.useFakeTimers();

		spy = vi.spyOn(console, 'log').mockImplementation(() => {});
	});

	afterEach(() => {
		spy.mockRestore();

		vi.useRealTimers();
	});

	it('errors if applied to a class', () => {
		expect(() => {
			// @ts-expect-error Allow decorator here
			@Throttle(100)
			class TestClass {}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors if applied to a property', () => {
		expect(
			() =>
				class TestProp {
					// @ts-expect-error Allow decorator here
					@Throttle(100)
					value = 123;
				},
		).toThrowErrorMatchingSnapshot();
	});

	it('logs once within the delay', () => {
		const test = new Test();

		test.log();

		expect(spy).toHaveBeenCalledWith('Logged!');

		test.log();
		test.log();
		test.log();

		expect(spy).toHaveBeenCalledTimes(1);

		vi.advanceTimersByTime(150);

		test.log();

		expect(spy).toHaveBeenCalledWith('Logged!');
		expect(spy).toHaveBeenCalledTimes(2);
	});

	it('passes arguments through', () => {
		const test = new Test();

		test.log(1, 2);
		test.log(3, 4);
		test.log(5, 6);

		vi.advanceTimersByTime(150);

		expect(spy).toHaveBeenCalledWith('Logged!', 1, 2);
		expect(spy).toHaveBeenCalledTimes(1);
	});
});
