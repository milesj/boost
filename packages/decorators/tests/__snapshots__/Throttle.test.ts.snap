// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Throttle() errors if applied to a class 1`] = `"\`@Throttle\` may only be applied to class methods."`;

exports[`@Throttle() errors if applied to a property 1`] = `"\`@Throttle\` may only be applied to class methods."`;
